### ISSUE TITLE
`$PATH` only present when Atom is launched from command line

### ISSUE THREAD

**Issue by dmnd** [AtomLinter/Linter#150]
When Atom is launched from the command line with `atom`, it has access to `$PATH` as defined by the shell so the linter can often find executables.

But if Atom is launched via the dock, Spotlight, Applications folder, or other GUI method, it does not inherit `$PATH`. This means that linters will fail to find their executables. I bet this problem is behind a lot of the bug reports that see the `spawn ENOENT` message.

**Comment by DavidAwad** [comment]
Confirmed as well, Ubuntu 14.10 on MacBook Air 2013

**Comment by adampats** [comment]
Same issue for me.  Launching from command line works fine (linter-rubocop loads and functions).  However, launching from dock or via Finder file association fails to find rubocop and throws this error:

    /Applications/Atom.app/Contents/Resources/app/src/buffered-process.js:
    104 Uncaught BufferedProcessError: Failed to spawn command `rubocop`. Make sure `rubocop` is installed and on your PATH

Mavericks 10.9.5 - Atom 0.169.0

**Comment by suhz** [comment]
I'am having the same problem. Using OS X Yosemite. Atom 0.170.

`
Failed to spawn command luac. Make sure luac is installed and on your PATH
`

**Comment by mjgold** [comment]
I'm having this problem, too, when launching Atom from the Mac doc:  Uncaught BufferedProcessError: Failed to spawn command `rubocop`. Make sure `rubocop` is installed and on your PATH
/Applications/Atom.app/Contents/Resources/app/src/buffered-process.js:100 Uncaught BufferedProcessError: Failed to spawn command `rubocop`. Make sure `rubocop` is installed and on your PATH

Launching from Terminal loads rubocop correctly.

**Comment by andersennl** [comment]
+1, Atom Version 0.175.0 OSX Yosemite.

**Comment by soulwax** [comment]
+1, Atom 0.176.0 Win 10 (should't be a problem), pylint path works from the cmd

**Comment by dpflucas** [comment]
+1, launching from terminal. Atom Version 0.176.0 (0.176.0), OSX Mavericks.

```
Failed to spawn command `rubocop`. Make sure `rubocop` is installed and on your PATH
```

**Comment by beck** [comment]
Would configuration be an appropriate solution?  Allow users to configure linter with a list of bin dirs to append to the PATH.

I recently tried to solve this issue myself system-wide and discovered that OSX 10.10 has made it increasingly difficult to configure your system path for GUI apps.

**Comment by dmnd** [comment]
@beck Maybe we could use `bash --login -c "echo $PATH"` then prepend it to the linter command as suggested by @bolinfest. I'm not sure if `path_helper` is a better option on OS X.

I haven't fixed this because honestly I never run into it. I always start Atom from the Terminal.

**Comment by MatthewHerbst** [comment]
@dmnd I think most many early adopters will start Atom from command line. At the same time, this needs to get fixed so that when Atom is more developed, less technical people can use it no problem. Also, from my personal experience, I have Atom set up as my default text editor, so I often open Atom by double clicking on a file on my desktop. 

@beck As per the above, I think users having to configure directories is too complex as a task for the eventual broad audience that we want Atom to have.

**Comment by dmnd** [comment]
@MatthewHerbst sorry I didn't mean to suggest this shouldn't be fixed. I'm just selfish when deciding where to spend my own time :smile: 

I would definitely merge something that fixes this.

**Comment by beck** [comment]
>  users having to configure directories is too complex as a task for the eventual broad audience that we want Atom to have

Agreed.

> use `bash --login -c "echo $PATH"`

I like this approach, should work for *nix

Do we care / how does one fetch window's `%PATH%`?

**Comment by beck** [comment]
Side note, the bash login cmd can be fickle if any of the PATH is set using an alias.

Simple example:
```
alias xcode_path="xcode-select --print-path"
XCODE=`xcode_path`/usr/bin
export PATH="$PATH:$XCODE"
```

Then you'll see:
```
$ bash --login -c "echo $PATH"
/Users/doug/.bash_profile: line 114: xcode_path: command not found
/usr/local/bin:/Users/doug/bin:/Applications/Xcode.app/Contents/Developer/usr/bin:....
```

Even so, it is still useable and better than a manual config.

**Comment by dmnd** [comment]
> How would this work on windows?

Not sure. But I think `%PATH%` (or whatever it is) on Windows might be available to GUI apps anyway so it mightn't even suffer from this problem. I'm just guessing though.

**Comment by reuwsaat** [comment]
"Failed to spawn command gjslint. Make sure gjslint is installed and on your PATH" on ubuntu 14.04

i get this with "linter-gjslint" but not "linter-jshint"

**Comment by beck** [comment]
> i get this with "linter-gjslint" but not "linter-jshint"

That's because the jshint linter [has extra logic to find the node bin](https://github.com/AtomLinter/linter-jshint/blob/master/lib/init.coffee#L5) whereas the gjslint linter does not.

**Comment by Ryan050** [comment]
+1, but launching from terminal works fine

**Comment by andronat** [comment]
From the terminal is working fine (Max OS X 10.9.5). But not when using the Icon from Dock. Is this a problem of linter or atom himself??

**Comment by bugaiov** [comment]
+1

**Comment by jmecom** [comment]
+1, Atom 0.180.0, OS X Yosemite. Also fails when launching from terminal as well.

**Comment by andronat** [comment]
Atom 0.181.0, OS X Yosemite. Also fails.

**Comment by jtkiley** [comment]
+1 Atom 0.181.0, OS X 10.10.

Where is node getting the path from? It's missing one of the paths in `/etc/paths`, and it doesn't pick up anything in `/etc/paths.d/`, in addition to not getting anything exported in `.bash_profile`. I'm checking by looking at the object returned by `process.env` in the error console.

**Comment by jtkiley** [comment]
It's ugly, but this (added to `init.coffee`) is how I'm getting it to find flake8:

~~~coffee
process.env.PATH = ["/Library/Frameworks/Python.framework/Versions/3.4/bin",
    process.env.PATH].join(":")

~~~

**Comment by patmcd** [comment]
+1.
Atom 0.182.0 
OSX 10.10.2

even occurs when launching from terminal.

**Comment by rothnic** [comment]
Thanks to @jtkiley, this is how I got things working with anaconda on osx:

```coffee
process.env.PATH = ["/Users/rothnic/anaconda/bin", process.env.PATH].join(":")
```

**Comment by crystoll** [comment]
Can verify, I had same issue with gjslint, I added /usr/bin and /usr/local/bin to path in init.coffee script under ~/.atom - and now no errors, it works:

process.env.PATH = ["/usr/bin","/usr/local/bin",
    process.env.PATH].join(":")

**Comment by henridwyer** [comment]
I'm getting a similar error:
``` Uncaught BufferedProcessError: Failed to spawn command `/Applications/Atom.app/Contents/Resources/app/apm/bin/node`. Make sure `/Applications/Atom.app/Contents/Resources/app/apm/bin/node` is installed and on your PATH ``` 
on Mac OS 10.10.2, since Atom version 0.186, when I launch from command line and from the dock. 

node is on my path, at /usr/local/bin, but can also be found at that location.

**Comment by paulbaker3** [comment]
I was getting a similar error with rubocop: `failed to spawn rubocop make sure rubocop is installed and on your path`

@crystoll's work around worked for me. Here's my ~/.atom/init.coffee
```
process.env.PATH = ["/usr/bin",
                    "/usr/local/bin",
                    "/Users/paulbaker/.rbenv/shims",
                    "/Users/paulbaker/.rbenv/bin"
process.env.PATH].join(":")
```

**Comment by jjoedouglas** [comment]
@paulbaker3 's ~/.atom/init.coffee worked for me as well.

Here is a more minimal version (with typo correction) which worked for me:

```
process.env.PATH = ["/usr/bin",
                    "/usr/local/bin",
                    ].join(":")
```

**Comment by MatthewHerbst** [comment]
Would something along the lines of the answer provided [here](http://apple.stackexchange.com/questions/106355/setting-the-system-wide-path-environment-variable-in-mavericks) work to help alleviate this? The downside is that it would require maintaining each time a plugin is added, updated, or removed.

Further, in [Why did my TeX-related GUI program stop working in Mac OS X Yosemite?](http://tex.stackexchange.com/questions/208181/why-did-my-tex-related-gui-program-stop-working-in-mac-os-x-yosemite), there is some background on a further related issue, but what also seems to be some solutions to fix. However, the author warns strongly against the first solution I posted since apparently it changes variables for all running programs.

**Comment by julienpa** [comment]
Is something planned to fix it someday? I mean, this issue is quite old but the problem seems unchanged; and in some cases (like mine) it makes the software simply unusable (I need linters but most of them throw errors: `Failed to spawn command scss-lint. Make sure scss-lint is installed and on your PATH`)
I tried everything mentioned here but nothing fixed the problem, not even launching Atom from the command line. I am on OS X 10.10.2 with Atom 0.187.0.
I'll use another editor in the meantime, but it is a sad and radical solution to an apparently small problem.
Thanks for any insight on that!

**Comment by MatthewHerbst** [comment]
@julienpa Based on the above post of mine, it seams this is more a bug in OSX than in Atom, though there are certainly some potential hacks around the bug. Not sure why it won't work when you start from command line - that seems to fix it for everyone else.

**Comment by julienpa** [comment]
@MatthewHerbst ah, indeed I had missed the "Yosemite issue" part. Anyway, I was already using some linters on Yosemite before and it was working fine. But I recently added rubocop, scss-lint and haml-lint and none of them work; all of them throw the same errors.
I tried launching Atom with the command line from the `~` directory as well as various projects' directories. Same error all the time.
Thanks for the reply.

**Comment by jtkiley** [comment]
@julienpa Have you tried adding to the path in `init.coffee` like my example above? You'll need the path to the linters you want to use, of course, but it should work.

I agree that it should be a high priority to fix. Linters are really handy, and it would be nice not to have a hurdle for people to use them.

**Comment by julienpa** [comment]
@jtkiley yes I tried that but it didn't work. Unfortunately I didn't have time to look more deeply into it, and I switched to Sublime Text in order to get the work done without the tons of warnings (I stopped disabling/uninstalling packages in the hope that it would miraculously be fixed one day).
Thanks!

**Comment by gabll** [comment]
+1. I've tried to modify the path in init.coffee, open atom from terminal and from finder, but it's still not working. I'm out of ideas

**Comment by laurelnaiad** [comment]
What does it mean if you don't even have ~/.atom/init.coffee at all?

```
$ ls -la ~/.atom
total 32
drwxr-xr-x   10 dmaddox  wheel   340 May 16  2014 .
drwxr-xr-x+  93 dmaddox  wheel  3162 Mar 13 08:57 ..
drwxr-xr-x    4 dmaddox  wheel   136 May 16  2014 .node-gyp
drwxr-xr-x    6 dmaddox  wheel   204 Feb  5 09:39 compile-cache
-rw-r--r--    1 dmaddox  wheel   798 Mar 13 09:08 config.cson
-rw-r--r--    1 dmaddox  wheel  1021 Aug 22  2014 keymap.cson
drwxr-xr-x   32 dmaddox  wheel  1088 Mar 13 09:03 packages
-rw-r--r--    1 dmaddox  wheel   415 Mar 29  2014 snippets.cson
drwxr-xr-x  146 dmaddox  wheel  4964 Mar 12 18:15 storage
-rw-r--r--    1 dmaddox  wheel  1227 Dec  9 18:53 styles.less
```

And wouldn't it be great if atom itself would do the work (once) to find the node binary and provide a property to its extensions where they can grab it?

My symptoms:

```
Failed to spawn command /Applications/Atom.app/Contents/Resources/app/apm/bin/node. Make sure /Applications/Atom.app/Contents/Resources/app/apm/bin/node is installed and on your PATH
```

Is see individual linters have init.coffee, but nothing is out in ~/.atom.

**Comment by julienpa** [comment]
@gabll @laurelnaiad: yeah... I had to switch back to Sublime Text (which I didn't want to do). There, with a _Sublime-Linter_ package, you just need to install specific linters with npm or gem and done! It's working.

**Comment by opennomad** [comment]
I've been running just fine for a couple of months now until this morning. I tried various combinations of running from the terminal and setting `PATH` as suggested above. None of these items fixed my issue.

So I moved my `~/.atom` out of the way and started fresh. Installing almost all of the same packages via `apm` and my linters for `shellcheck` and `ruby` are once again working perfectly.

Assuming that it wasn't anything odd I had installed, I moved the new `.atom` out of the way and put the old one back and moved the `compile-cache` out of the way. Again my linters are working.

Then for one more try I put the original `compile-cache` back and things are still working.

Sure feels like a gremlin in there somewhere. At any rate I'm back up and running and perhaps some of my shenanigans might help others.

**Comment by RodrigoJimmy** [comment]
As jjoedouglas said, just edit `~/.atom/init.coffee`.
Go to menu `Atom > Open Your Init Script`, just add this lines at the end of the file:

```
process.env.PATH = ["/usr/bin",
                    "/usr/local/bin",
                    ].join(":")
```

Works for me.

**Comment by tr33m4n** [comment]
Just FYI, under Linux/Ubuntu I had to add `/bin` to the path:
```
process.env.PATH = ["/bin",
                                   "/usr/bin",
                                   "/usr/local/bin",
                                   ].join(":")
```
Otherwise certain elements of the settings page wouldn't work correctly, such as searching for a package to install (would throw an error saying it could not find uname)

Cheers
Dan

**Comment by Trefex** [comment]
Adding 

```
process.env.PATH = ["/usr/bin",
                    "/usr/local/bin",
                    ].join(":")
```

to `.atom/init.coffee` fixed the issue for me.

:+1:

**Comment by bee-git** [comment]
Using Atom 0.198.0 on Windows 8.1 Professional together with XAMPP.  Trying to use linter-php.

Have entered the following line at the end of .atom/init.coffee

    process.env.PATH = ["C:/PortableApps/XAMPP/App/xampp/php", process.env.PATH].join(":")

With this entered, I nevertheless get this error message when entering code:

    'php' could not be spawned. Is it installed and on your path? 
    If so please open an issue on the package spawning the process.

with a "Uncaught Error: spawn php ENOENT" in the log file.

I tried adding the following lines to the .atom/config.cson file:

    "linter-php":
        phpExecutablePath: "C:/PortableApps/XAMPP/App/xampp/php/"

in addition to the entry in the .atom/init.coffee file, but that didn't eliminate the problem.

In other words, the solutions that have been suggested thus far in this (otherwise helpful) thread don't seem to work in a Windows 8.1 environment.

**Comment by Ricain** [comment]
Path should contain only directory, right? not a binary like php, so try this:

```py
process.env.PATH = ["C:/PortableApps/XAMPP/App/xampp/", process.env.PATH].join(":")
```

**Comment by bee-git** [comment]
@Ricain - Thank you for the suggestion. However, ```C:/PortableApps/XAMPP/App/xampp/php/``` is, in fact, the directory where the php executable ```php.exe``` is located.  That is, the fully specified path and file name of the php executable is 

    C:/PortableApps/XAMPP/App/xampp/php/php.exe

As I mentioned in the update to my original post, the problem can be solved  by placing the path for the php executable in the system ```PATH``` variable.  But that really shouldn't be necessary given the configuration and init script options that are supposed to allow you to specify where the executable is located.

**Comment by beck** [comment]
Has anyone else noticed that when you checkout a different branch (in a project dir already opened in atom), the path is destroyed? Thus in turn, breaking linter.

**Comment by ChrisAndrew** [comment]
Hey guys,

    process.env.PATH = ["/usr/bin",
                        "/usr/local/bin"].join(":")

also tried

    process.env.PATH = ["/usr/bin",
                        "/usr/local/bin",
                        process.env.PATH].join(":")

to my ~/.atom/init.coffee didn't work. I also noted that Atom when started from command line works fine but didn't when started from my Xubuntu panel, so I edited the Atom shortcut (laucher) in my panel and changed the command from */opt/atom/atom %F* to *bash --login -c "atom"*, as suggested above.

Now all my ruby related packages are working!

**Comment by DavidAwad** [comment]
@Trefex THANK YOUUUU. YOU HAVE SOLVED MY LIFELONG STRUGGLE WITH THIS EDITOR.

**Comment by ssfrr** [comment]
It seems that `init.coffee` isn't run until after packages are loaded, so if one of your packages tries to spawn a shell command that's not on your default `$PATH` then it still doesn't work. As fas as I can tell the only alternative in that case is tweaking the launcher to run Atom from within a shell environment.

**Comment by rennsport** [comment]
I'm running OS X 10.10.5 and Atom 1.0.9 and I'm having this issue both launching from GUI and terminal. Are there any other solutions other than editing the `init.coffee` because that doesn't seem to work for me.

**Comment by steelbrain** [comment]
@rennsport why not specify a path to the executable in linter config?

**Comment by rennsport** [comment]
@steelbrain So my linter directory is /Users/[user]/.atom/packages/linter what file would I need to edit because none are standing directly out as a config file.

**Comment by steelbrain** [comment]
@rennsport You won't need to edit any file for this, just goto settings view, the linter provider you want to specify executable for, and there should be a field for it.

**Comment by rennsport** [comment]
@steelbrain My [linter provider](https://github.com/atom-community/linter) doesn't seem to have those options in settings

**Comment by steelbrain** [comment]
@rennsport that is the base linter you are pointing at, by a "provider", I meant packages like "linter-jshint", "linter-eslint" and "linter-flake8"

**Comment by rennsport** [comment]
@steelbrain I feel slightly less capable now as I linked this project within the project itself without realizing it haha. Sadly however, the actual [linter provider](https://github.com/AtomLinter/linter-pylint) does not have such settings unless I'm missing something obvious again.

**Comment by steelbrain** [comment]
@rennsport It does have that setting and it's called `executable`, see https://github.com/AtomLinter/linter-pylint/blob/master/lib/main.coffee#L9
but if you are unable to see it for some reasons, then I would suggest you try this and let me know the results of it
```
mv ~/.atom ~/.atom-bkp
apm install linter
apm install linter-pylint
```

**Comment by rennsport** [comment]
@steelbrain I now see said settings. Maybe the theme I was using hid them? Anyways I have now gotten around the first error, and I have a second (I apologize for my novice understand of this). Error: "Traceback (most recent call last):" I can show the full error in a pastebin or something similar.

**Comment by steelbrain** [comment]
@rennsport can we please chat in the atom slack or at least in a separate issue, I have to open up this super-long page each time and tons of people are notified each time via email/push.

**Comment by anildigital** [comment]
In linter-rubycop plugin,

I have rubocop command set like this.

```
/Users/anil/.gem/ruby/2.2.4/bin/rubocop
```

But when I open Ruby project from Atom menu `File -> Open..`, it gives me following error

```
Error: /Users/anil/.rubies/ruby-2.2.4/lib/ruby/2.2.0/rubygems/dependency.rb:315:in to_specs': Could not find 'rubocop' (>= 0) among 229 total gem(s) (Gem::LoadError) Checked in 'GEM_PATH=/Users/anil/.gem/ruby/2.2.0:/Users/anil/.rubies/ruby-2.2.4/lib/ruby/gems/2.2.0', executegem envfor more information from /Users/anil/.rubies/ruby-
```

But when I open project from command like like `atom .` I don't see this error.

My Atom version is 1.3.2 and I am using El Capitan.

**Comment by Arcanemagus** [comment]
This is an issue that is planning on being solved in Atom (see https://github.com/atom/atom/issues/4126), with a temporary workaround planned in https://github.com/AtomLinter/atom-linter/issues/62 for linter plugins that utilize it. For any that don't they will need to work around this limitation on their own.

I'm locking this issue as it hasn't been the proper repository for this issue for a _very_ long time in the first place, and any further work on resolving the issues presented here will be done in the linked issues.

For any users on OS X until a workaround for Apple's "interesting" decision is put in place, the simplest method is to launch Atom from a terminal that has a full environment set up using `atom .`.

**Comment by steelbrain** [comment]
Here's an announcement for you, we have recently published a release of `atom-linter` (npm helper module for linter providers) that now automatically determines `$PATH` even if atom is not launched from terminal. Therefore all linters using it for execution will be working consistently from both dock and cli
